x-pack/plugin/security/qa/jwt-realm/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRestIT.java

```diff
@@ -0,0 +1,458 @@
+        final String principal = randomPrincipal();
+        final List<String> roles = randomRoles();
+        final String randomMetadata = randomAlphaOfLengthBetween(6, 18);
+        createUser(principal, roles, Map.of("test_key", randomMetadata));
+
+        try {
+            final SignedJWT jwt = buildAndSignJwtForRealm2(principal);
+            final TestSecurityClient client = getSecurityClient(jwt, VALID_SHARED_SECRET);
+
+            final Map<String, Object> response = client.authenticate();
+
+            assertThat(response.get(User.Fields.USERNAME.getPreferredName()), is(principal));
+            assertThat(assertMap(response, User.Fields.AUTHENTICATION_REALM), hasEntry(User.Fields.REALM_NAME.getPreferredName(), "jwt2"));
+            assertThat(assertList(response, User.Fields.ROLES), Matchers.containsInAnyOrder(roles.toArray(String[]::new)));
+            assertThat(assertMap(response, User.Fields.METADATA), hasEntry("test_key", randomMetadata));
+
+            // The user has no real role (we never define them) so everything they try to do will be FORBIDDEN
+            final ResponseException exception = expectThrows(
+                ResponseException.class,
+                () -> client.getRoleDescriptors(new String[] { "*" })
+            );
+            assertThat(exception.getResponse(), hasStatusCode(RestStatus.FORBIDDEN));
+        } finally {
+            deleteUser(principal);
+        }
+    }
+
+    public void testFailureOnInvalidHMACSignature() throws Exception {
+        final String principal = randomPrincipal();
+        final List<String> roles = randomRoles();
+        try {
+            createUser(principal, roles, Map.of());
+
+            final JWTClaimsSet claimsSet = buildJwtForRealm2(principal, Instant.now());
+
+            {
+                // This is the correct HMAC passphrase (from build.gradle)
+                final SignedJWT jwt = signHmacJwt(claimsSet, "test-HMAC/secret passphrase-value");
+                final TestSecurityClient client = getSecurityClient(jwt, VALID_SHARED_SECRET);
+                assertThat(client.authenticate(), hasEntry(User.Fields.USERNAME.getPreferredName(), principal));
+            }
+            {
+                // This is not the correct HMAC passphrase
+                final SignedJWT invalidJwt = signHmacJwt(claimsSet, "invalid-HMAC-passphrase-" + randomAlphaOfLength(12));
+                final TestSecurityClient client = getSecurityClient(invalidJwt, VALID_SHARED_SECRET);
+                // This fails because the HMAC is wrong
+                final ResponseException exception = expectThrows(ResponseException.class, client::authenticate);
+                assertThat(exception.getResponse(), hasStatusCode(RestStatus.UNAUTHORIZED));
+            }
+        } finally {
+            deleteUser(principal);
+        }
+
+    }
+
+    public void testAuthenticationFailureIfDelegatedAuthorizationFails() throws Exception {
```

Comment:
Non-issue: Here is a general question I was wondering about. It is not specific to JWT or any other realm.

Why does Realm.authenticate() perform authc and authz? If authc succeeds and authz fails, is that equivalent to successful authentication with no roles?